src/transform/src/column_knowledge.rs

```diff
@@ -475,13 +475,17 @@ pub fn optimize(
                 } => {
                     let knowledge2 = knowledge_stack.pop().unwrap();
                     let knowledge1 = knowledge_stack.pop().unwrap();
-                    if knowledge1.value.is_some() && knowledge2.value.is_some() {
+                    if knowledge1.value.is_some() || knowledge2.value.is_some() {
                         e.reduce(column_types);
                     }
                     DatumKnowledge::from(&*e)
                 }
                 MirScalarExpr::CallVariadic { func: _, exprs } => {
-                    if (0..exprs.len()).all(|_| knowledge_stack.pop().unwrap().value.is_some()) {
+                    let mut knows = Vec::new();
+                    for _ in exprs {
+                        knows.push(knowledge_stack.pop().unwrap());
+                    }
+                    if knows.iter().any(|k| k.value.is_some()) {
```

Comment:
I might be missing something in this logic, but I *think* it could be `knowledge_stack.drain(0 .. exprs.len())...`? At least, afaict the only thing that is happening is that we are unconditionally draining `knowledge_stack` and using an `any` instead (which risks short circuiting with the above logic).

Reply:
I added a comment.